### DISCUSSION TITLE
IPPO configuration settings

### DISCUSSION THREAD

**Discussion by JulienHansen** [Toni-SM/skrl#270] [Q&A] [answered]
Hi @Toni-SM ,

I believe I don't fully understand the model instantiator. I used the following IPPO YAML file for my application, which involves two drones—one attacking and the other defending. What I can't comprehend is that when testing my environment and allowing the attacker to reach a desired position with the following reward function:

```py
    def _get_rewards(self) -> dict[str, torch.Tensor]:
        rewards = {}
        for agent in self.cfg.possible_agents:
            if agent == "drone_attack":
                lin_vel = torch.sum(torch.square(self._drone_attack.data.root_lin_vel_b), dim=1)
                ang_vel = torch.sum(torch.square(self._drone_attack.data.root_ang_vel_b), dim=1)
                distance_to_goal = torch.linalg.norm(self._desired_pos_w - self._drone_attack.data.root_pos_w, dim=1)
                distance_to_goal_mapped = 1 - torch.tanh(distance_to_goal / 0.8)
                reward = (
                    lin_vel * self.cfg.lin_vel_reward_scale +
                    ang_vel * self.cfg.ang_vel_reward_scale +
                    distance_to_goal_mapped * self.cfg.distance_to_goal_reward_scale
                ) * self.step_dt
            elif agent == "drone_defense":
                # Implement the defender's reward logic here
                pass
            else:
                raise ValueError(f"Unknown agent key: {agent}")
            
            rewards[agent] = reward
            # Logging
            #self._episode_sums[f"{agent}_reward"] += reward

        return rewards
```

the one defending learn to fly even though with IPPO i would have though it would only stay on the ground as it doesn't have any reward, here is my IPPO config, am i missing something ? 

```yaml 

seed: 42


# Models are instantiated using skrl's model instantiator utility
# https://skrl.readthedocs.io/en/latest/api/utils/model_instantiators.html
models:
  separate: True
  policy:  # see gaussian_model parameters
    class: GaussianMixin
    clip_actions: False
    clip_log_std: True
    min_log_std: -20.0
    max_log_std: 2.0
    initial_log_std: 0.0
    network:
      - name: net
        input: STATES
        layers: [512, 256, 128]
        activations: elu
    output: ACTIONS
  value:  # see deterministic_model parameters
    class: DeterministicMixin
    clip_actions: False
    network:
      - name: net
        input: STATES
        layers: [512, 256, 128]
        activations: elu
    output: ONE


# Rollout memory
# https://skrl.readthedocs.io/en/latest/api/memories/random.html
memory:
  class: RandomMemory
  memory_size: -1  # automatically determined (same as agent:rollouts)


# IPPO agent configuration (field names are from IPPO_DEFAULT_CONFIG)
# https://skrl.readthedocs.io/en/latest/api/multi_agents/ippo.html
agent:
  class: IPPO
  rollouts: 16
  learning_epochs: 5
  mini_batches: 4
  discount_factor: 0.99
  lambda: 0.95
  learning_rate: 5.0e-04
  learning_rate_scheduler: KLAdaptiveLR
  learning_rate_scheduler_kwargs:
    kl_threshold: 0.016
  state_preprocessor: RunningStandardScaler
  state_preprocessor_kwargs: null
  value_preprocessor: RunningStandardScaler
  value_preprocessor_kwargs: null
  random_timesteps: 0
  learning_starts: 0
  grad_norm_clip: 1.0
  ratio_clip: 0.2
  value_clip: 0.2
  clip_predicted_values: True
  entropy_loss_scale: 0.0
  value_loss_scale: 2.0
  kl_threshold: 0.0
  rewards_shaper_scale: 1.0
  time_limit_bootstrap: False
  # logging and checkpoint
  experiment:
    directory: "quadcopter_direct"
    experiment_name: "100k"
    write_interval: auto
    checkpoint_interval: auto
    wandb: True
    wandb_kwargs:



# Sequential trainer
# https://skrl.readthedocs.io/en/latest/api/trainers/sequential.html
trainer:
  class: SequentialTrainer
  timesteps: 36000
  environment_info: log

```

Thanks in advance

**Comment by JulienHansen** [comment]
Is there a way in the config file to ensure that both agents have distinct network ?

**Reply by JulienHansen** [reply]
I'm not sure but could the unwanted behavior of my drone come from a shared buffer in the memory logic ? I have check your IPPO code, is it possible it could come from that ? If yes, how could i change that to ensure nothing is shared among agents
Thanks

**Comment by Toni-SM** [comment]
Hi @JulienHansen 

> Is there a way in the config file to ensure that both agents have distinct network?

Currently no in the Runner/`.yaml` workflow integrated in Isaac Lab. I am working to allow it.
Meanwhile, you can use skrl's standalone scripts (see [skrl's NVIDIA Isaac Lab](https://skrl.readthedocs.io/en/latest/intro/examples.html#nvidia-isaac-lab) examples, and some [multi-agent scripts examples](https://skrl.readthedocs.io/en/latest/intro/examples.html#bi-dexhands-environments)) to have full control on the definition of the model for each agent.

If you are for it, try to create the script for your training... I can help you with defining it if there is anything that is not clear

**Reply by JulienHansen** [reply]
Thanks for the fast answer i will try to define it, i had another question has i'm not completely sure of the source of my problem. Here is the situation, i have 2 drone and one of them has reward all equal to 0 and the other has logic reward with IPPO i would expect the first drone to not learn anything at all, but in practice it learns to fly like the other. How is this behavior achieve ? Does it come from some king of shared memory ? shared weight among the model ? I checked the code and it doesn't seem so regarding the model , even though every agents needs to have the same architecture , the models seem independent , however i didn't fully understand the memory part so maybe the agent learns thanks to the random sampling in the episodes ?

**Reply by Toni-SM** [reply]
Hi @JulienHansen 

I see...
In your task, you need to also specify the reward for the second agent, otherwise second agent will be assigned same reward value computed for the first agent 

```python
    ...
           elif agent == "drone_defense":
                # Implement the defender's reward logic here
                reward = torch.zeros((self.num_envs), self.device)
    ...
    rewards[agent] = reward
```

**Reply by JulienHansen** [reply]
Yeah sorry i forgot to update my code snippet i specify the reward i checked that drone_defense reward was 0 afterwards and it was but it's still learning to fly , i also try to put all it's observations to 0 without any change

**Reply by Toni-SM** [reply]
And `separate: True` (in agent config), right?

**Reply by JulienHansen** [reply]
Yes separate: True , reward to 0, observation to 0 and still it learns to fly

**Reply by Toni-SM** [reply]
Mmmm, this should not be possible under these conditions for sure.
Could it be possible to share more code (repo link to the task and configs)?

**Reply by JulienHansen** [reply]
Here is a link to my environment:(https://github.com/JulienHansen/Isaac_drone/blob/main/source/Isaac_drone/Isaac_drone/tasks/Isaac_drone_v3/drone_defense.py) the runner is under the agent directory , you can see below with the video both drone learn to fly after 2 to 3 minutes even though the observation and reward returns are : 

```yaml
observations {'drone_attack': tensor([[ 81.9614, -74.1741,   0.2507,  ...,   0.5878,  -0.6304,   0.5071],
        [ 84.6072, -69.0299,   0.8075,  ...,   0.8232,  -0.1046,   0.5580],
        [ 75.7429, -65.0519,   0.6794,  ...,   0.4354,  -0.3997,   0.8066],
        ...,
        [-84.8021, -52.3664,   0.1376,  ...,  -0.7545,   0.5468,   0.3630],
        [-84.4955, -46.3951,   0.8135,  ...,   0.0898,   0.1442,  -0.9855],
        [-75.1423, -43.0596,   0.7694,  ...,   0.6681,   0.3957,  -0.6301]],
       device='cuda:0'), 'drone_defense': tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0')}
```

Reward (just a part of it) : 

```yaml 
      -1.4700e-01, -6.9017e-02, -5.3439e-02, -1.5111e-01, -2.4640e-02,
        -4.5588e-02, -3.7796e-01, -1.5694e-01, -5.4094e-02, -1.4780e-01,
        -2.7674e-01, -3.4840e-02, -5.3781e-02, -2.4543e-02, -5.1207e-01,
        -3.7217e-01, -1.0066e-01, -7.3181e-01, -1.7562e-01, -1.2206e-01,
        -9.6783e-02, -2.8401e-01, -4.3191e-02, -3.3995e-02, -8.3629e-01,
        -6.2876e-02, -3.0556e-01, -1.5836e-01, -1.6242e-01, -1.6466e-01,
        -4.0862e-01, -7.4551e-02, -2.4580e-01, -1.9595e-02, -3.8499e-02],
       device='cuda:0'), 'drone_defense': tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
```


I'm sure i'm doing something wrong or that i miss something in either the config or the agents design...
Thanks for all the help/informations you already provided me.






[Screencast from 2025-02-14 09-27-20.webm](https://github.com/user-attachments/assets/7cd19707-be73-4479-8635-e802976f63e0)

**Reply by JulienHansen** [reply, marked as answer]
Okay sorry i finally figured it out, it comes from my side. In the apply_actions i had an error with a loop iterator.
Thank you so much for the time you took to response !

**Reply by Toni-SM** [reply]
Glad to hear it works!

**Comment by JulienHansen** [comment]
Hi @Toni-SM, 
I just have another question with IPPO is it possible to use composite observation_space in IsaacLab for multi-agent task ? 
Something similair to this but for each agent : 

```py
    observation_space = {
    "robot-state": 12,
    "raycaster": [1, 36, 4],
    }
```

**Reply by Toni-SM** [reply]
In principle yes, skrl's [model instantiator](https://skrl.readthedocs.io/en/latest/api/utils/model_instantiators.html) used by skrl's [runners](https://skrl.readthedocs.io/en/latest/api/utils/runner.html) fully operates with the original spaces types.

Now, if you are using or are planning to use skrl's standalone scripts (as shown in [skrl's Examples](https://skrl.readthedocs.io/en/latest/intro/examples.html)), you need to use [spaces utilities](https://skrl.readthedocs.io/en/latest/api/utils/spaces.html) manually when defining the models.
